Boxes with `position: fixed` should not be a part of scrollable overflow of viewport

longvatrong111 opened the pull request:
<!-- Please describe your changes on the following line: -->
When calculate `scrollable_overflow` rect for Fragment tree, should ignore child with `position: fixed`:
[Fixed boxes do not move when the document is scrolled](https://www.w3.org/TR/css-position-3/#fixed-position).

Test:
`tests/wpt/tests/css/css-masking/clip-path/clip-path-fixed-scroll.html`

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #35928 (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

longvatrong111 committed:
Boxes with position:fixed should not be a part of scrollable overflow of viewport

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -384,6 +385,15 @@ impl BoxTree {
             .fold(PhysicalRect::zero(), |acc, child| {
+                // https://www.w3.org/TR/css-position-3/#fixed-cb
+                // As a result, parts of fixed-positioned boxes that extend outside
+                // the layout viewport/page area cannot be scrolled to and will not print.
+                if let Fragment::Box(fragment) = child {
+                    if fragment.borrow().style.get_box().position == Position::Fixed {
+                        return acc;
+                    }
+                }
```
Should this also be the case for boxes with `Position::Absolute`?

longvatrong111 replied:
I agree. It's also confirmed by @Loirooriol in the issue ticket.

xiaochengh replied:
I don't think so. Check this test case in Chrome: https://jsfiddle.net/w1s7vhd4/

stevennovaryo replied:
www.w3.org/TR/css-overflow-3/#scrollable-overflow-region
> The border boxes of all boxes for which it is the containing block ...


By this definition, I suppose for the initial containing block, the scrollable overflow should consider absolute positioned boxes. Following that, I am not sure whether the calculation of scrollable overflow is correct for the children, because for a scroll container we should consider only the children in which the scroll container established a containing block on.

Kindly check this test too.
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Nested Elements</title>
    <style>
        .container {
            height: 500px;
            width: 300px;
            overflow: auto;
            border: 1px solid black;
            transform: translate(0px, 0px);
        }
        .fixed-element {
            position: fixed;
            bottom: -50px;
            right: -50px;
            height: 100px;
            width: 100px;
            background-color: lightcoral;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="fixed-element">
        </div>
    </div>
</body>
</html>
```

longvatrong111 replied:
@stevennovaryo Do you mean the parent-children relationship should reflect the containing block chain?

mrobinson commented on components/layout_2020/flow/root.rs:
```diff
@@ -384,6 +384,21 @@ impl BoxTree {
             .iter()
             .fold(PhysicalRect::zero(), |acc, child| {
+                // https://www.w3.org/TR/css-position-3/#fixed-cb
+                // As a result, parts of fixed-positioned boxes that extend outside
+                // the layout viewport/page area cannot be scrolled to and will not print.
```
The comment is a bit out of date here now because it only talks about fixed position elements.

Loirooriol reviewed (commented):
It would be good if you can add a test for `position: absolute` and which doesn't rely on css-masking thingies. But maybe #35019 should be addressed first.

longvatrong111 commented:
@Loirooriol I see, will wait for https://github.com/servo/servo/issues/35019 then add test cases later.

xiaochengh commented:
Btw, in Servo's implementation of the box tree, what's the parent node for a `position: fixed` element? Is it the root element, or do we have a separate ICB node?

I've encountered a similar issue in Blink, where the root LayoutObject (`LayoutView`) represents both the root element and the ICB. Then the handling of `position: fixed` becomes quite complicated.

Btw you shouldn't just check `position: fixed`. You also need to check if it's containing block chain ends at the ICB. It's because there are other cases where we can establish a containing block that's not ICB, and then make `position: fixed` indifferent from `position: absolute`. For example: https://jsfiddle.net/hbdzvncr/

```html
<div class="container">
  <div class="positioned fixed">FIXED</div>
  <div class="positioned absolute">ABSOLUTE</div>
</div>
<style>
.container {
  transform: scale(1);
  width: 400px;
  height: 400px;
  overflow: scroll;
  background: gray;
}

.positioned {
  width: 100px;
  height: 100px;
  background: green;
  bottom: -200px;
}

.fixed {
  position: fixed;
}

.absolute {
  position: absolute;
  left: 200px;
}
</style>
```

longvatrong111 commented:
@xiaochengh

> Btw, in Servo's implementation of the box tree, what's the parent node for a `position: fixed` element? Is it the root element, or do we have a separate ICB node?

In current implementation of Servo, Containing Block Chain is formed in `BoxTree::layout`.

Before traversing and layouting box, a `PositioningContext` is created. When a `position: fixed` or `position: absolute`  is encountered, a `HoistedAbsolutelyPositionedBox` is created and pushed into `PositioningContext`, and later is hoisted up in the tree.

Then `PositioningContext::layout_initial_containing_block_children` uses initial containing block to construct `Fragment` for root element and all `HoistedAbsolutelyPositionedBox` which are hoisted up to the root.

So I think the Containing Block Chain is correct according to [spec](https://www.w3.org/TR/css-position-3/#fixed-cb) except  the root for `position: fixed` is initial containing block, not the view port.

xiaochengh commented:
Let's try not to complicate things... The bottom line is:
- A `position: absolute` element should contribute to the scrollable overflow of the root scrollable
- A `position: fixed` element should *not* contribute to it

Test case (https://jsfiddle.net/odqbx3jf/):
```html
<style>
.target {
  width: 100px;
  height: 100px;
  background: green;
}
body {
  width: 90vw;
  height: 90vh;
}
</style>
<div class="target" style="position: fixed; right: -200px"></div>
<div class="target" style="position: absolute; bottom: -200px"></div>
```

The page should be vertically scrollable but horizontally unscrollable

longvatrong111 commented:
Offline discussed: the  solution is acceptable, need to add some tests.
